allow synthetic base code path as long as none of the aliased tensors are subclasses

bdhirsh opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/119755. We are prototyping if this situation is generally going to be excercised more heavily when tracing FSDP.

This slightly relaxes the assertion for the case in AOTAutograd when we experience "multiple graph inputs that were aliased, where at least one was mutated".

The old version of the restriction was that we would blanket ban this case, whenever we detected that any inputs/outputs to the graph are subclasses.

Instead, we allow this case, and only ban it when any of the **aliased graph inputs in question** are subclasses.

In the repro linked above, there is a DTensor graph input, but it was not actually mutated: there are two additional (plain) tensor inputs that are aliased and mutated, which is now allows.

A few things to note about this PR:

* re-computing accurate `ViewAndMutationMetadata` in this synthetic base case for subclasses is painful, so I opted not to do that for now and regenerate the metadata instead. This might slow compile times for models that hit this code path (although it won't be quite as bad now that we have fake tensor caching). If we see noticeably slow compiles we can revisit

* need to add tests

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120090
* __->__ #119824
* #120106

bdhirsh committed:
allow synthetic base code path as long as none of the aliased tensors are subclasses

[ghstack-poisoned]

ezyang commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -909,6 +891,15 @@ def _same_dtype_views(view1, view2):
+                    """\
+Encountered duplicate inputs that are mutated in the graph, but at least one of these
+inputs is a tensor subclass. This is not supported today. You can try to
+remove the aliasing yourself as a workaround, or otherwise file an issue on github."""
```
Better is to preemptively file a GH issue and link it here

ezyang reviewed (approved):
For testing, I'd like to see a test of the new case that now passes through successfully, and also a test that ensures that we don't blow up when a subclass aliases with regular tensor that mutates.

bdhirsh committed:
Update on "allow synthetic base code path as long as none of the aliased tensors are subclasses"

Fixes https://github.com/pytorch/pytorch/issues/119755. We are prototyping if this situation is generally going to be excercised more heavily when tracing FSDP.

This slightly relaxes the assertion for the case in AOTAutograd when we experience "multiple graph inputs that were aliased, where at least one was mutated".

The old version of the restriction was that we would blanket ban this case, whenever we detected that any inputs/outputs to the graph are subclasses.

Instead, we allow this case, and only ban it when any of the **aliased graph inputs in question** are subclasses.

In the repro linked above, there is a DTensor graph input, but it was not actually mutated: there are two additional (plain) tensor inputs that are aliased and mutated, which is now allows.

A few things to note about this PR:

* re-computing accurate `ViewAndMutationMetadata` in this synthetic base case for subclasses is painful, so I opted not to do that for now and regenerate the metadata instead. This might slow compile times for models that hit this code path (although it won't be quite as bad now that we have fake tensor caching). If we see noticeably slow compiles we can revisit

* need to add tests




[ghstack-poisoned]

bdhirsh committed:
Update on "allow synthetic base code path as long as none of the aliased tensors are subclasses"

Fixes https://github.com/pytorch/pytorch/issues/119755. We are prototyping if this situation is generally going to be excercised more heavily when tracing FSDP.

This slightly relaxes the assertion for the case in AOTAutograd when we experience "multiple graph inputs that were aliased, where at least one was mutated".

The old version of the restriction was that we would blanket ban this case, whenever we detected that any inputs/outputs to the graph are subclasses.

Instead, we allow this case, and only ban it when any of the **aliased graph inputs in question** are subclasses.

In the repro linked above, there is a DTensor graph input, but it was not actually mutated: there are two additional (plain) tensor inputs that are aliased and mutated, which is now allows.

A few things to note about this PR:

* re-computing accurate `ViewAndMutationMetadata` in this synthetic base case for subclasses is painful, so I opted not to do that for now and regenerate the metadata instead. This might slow compile times for models that hit this code path (although it won't be quite as bad now that we have fake tensor caching). If we see noticeably slow compiles we can revisit

* need to add tests




[ghstack-poisoned]
